python/pyiceberg/manifest.py

```diff
@@ -104,29 +104,100 @@ class ManifestEntry(IcebergBaseModel):
-    added_rows_count: Optional[int] = Field()
-    existing_rows_counts: Optional[int] = Field()
-    deleted_rows_count: Optional[int] = Field()
-    partitions: Optional[List[PartitionFieldSummary]] = Field()
-    key_metadata: Optional[bytes] = Field()
+MANIFEST_FILE_SCHEMA: Schema = Schema(
+    NestedField(500, "manifest_path", StringType(), required=True, doc="Location URI with FS scheme"),
+    NestedField(501, "manifest_length", LongType(), required=True),
+    NestedField(502, "partition_spec_id", IntegerType(), required=True),
+    NestedField(517, "content", IntegerType(), required=False),
+    NestedField(515, "sequence_number", LongType(), required=False),
+    NestedField(516, "min_sequence_number", LongType(), required=False),
+    NestedField(503, "added_snapshot_id", LongType(), required=False),
+    NestedField(504, "added_data_files_count", IntegerType(), required=False),
+    NestedField(505, "existing_data_files_count", IntegerType(), required=False),
+    NestedField(506, "deleted_data_files_count", IntegerType(), required=False),
+    NestedField(512, "added_rows_count", LongType(), required=False),
+    NestedField(513, "existing_rows_count", LongType(), required=False),
+    NestedField(514, "deleted_rows_count", LongType(), required=False),
+    NestedField(507, "partitions", ListType(508, PARTITION_FIELD_SUMMARY_TYPE, element_required=True), required=False),
+    NestedField(519, "key_metadata", BinaryType(), required=False),
+)
+
+MANIFEST_FILE_FIELD_NAMES: Dict[int, str] = {pos: field.name for pos, field in MANIFEST_FILE_SCHEMA.fields}
+IN_TRANSFORMS: Dict[int, Callable[[Any], Any]] = {
+    3: lambda id: ManifestContent.DATA if id == 0 else ManifestContent.DELETES
+}
+OUT_TRANSFORMS: Dict[int, Callable[[Any], Any]] = {
+    3: lambda content: content.id
+}
+
+
+class ManifestFile(StructProtocol):
+    manifest_path: str
+    manifest_length: int
+    partition_spec_id: int
+    content: ManifestContent
+    sequence_number: int
+    min_sequence_number: int
+    added_snapshot_id: Optional[int]
+    added_data_files_count: Optional[int]
+    existing_data_files_count: Optional[int]
+    deleted_data_files_count: Optional[int]
+    added_rows_count: Optional[int]
+    existing_rows_count: Optional[int]
+    deleted_rows_count: Optional[int]
+    partitions: Optional[List[PartitionFieldSummary]]
+    key_metadata: Optional[bytes]
+
+    _projection_positions: Dict[int, int]
+
+    def __init__(self, read_schema: Optional[Schema] = None):
+        if read_schema:
+            id_to_pos = {field.field_id: pos for pos, field in enumerate(MANIFEST_FILE_SCHEMA.fields)}
+            self._projection_positions = {pos: id_to_pos[field.field_id] for pos, field in enumerate(read_schema.fields)}
+        else:
+            # all fields are projected
+            self._projection_positions = {pos: pos for pos, field in MANIFEST_FILE_SCHEMA.fields}
+
+    @property
+    def content(self) -> ManifestContent:
```

Comment:
This one shadows the `content` field. I don't think this is necessary if we go with the `IN_TRANSFORM`